packages/scripts/config/webpack.config.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * External dependencies
+ */
+const { BundleAnalyzerPlugin } = require( 'webpack-bundle-analyzer' );
+const LiveReloadPlugin = require( 'webpack-livereload-plugin' );
+const path = require( 'path' );
+
+/**
+ * Internal dependencies
+ */
+const { camelCaseDash } = require( '../utils' );
+
+/**
+ * Converts @wordpress/* string request into request object.
+ *
+ * Note this isn't the same as camel case because of the
+ * way that numbers don't trigger the capitalized next letter.
+ *
+ * @example
+ * formatRequest( '@wordpress/api-fetch' );
+ * // { this: [ 'wp', 'apiFetch' ] }
+ * formatRequest( '@wordpress/i18n' );
+ * // { this: [ 'wp', 'i18n' ] }
+ *
+ * @param {string} request Request name from import statement.
+ * @return {Object} Request object formatted for further processing.
+ */
+const formatRequest = ( request ) => {
+	// '@wordpress/api-fetch' -> [ '@wordpress', 'api-fetch' ]
+	const [ , name ] = request.split( '/' );
+
+	// { this: [ 'wp', 'apiFetch' ] }
+	return {
+		this: [ 'wp', camelCaseDash( name ) ],
+	};
+};
+
+const wordpressExternals = ( context, request, callback ) => {
+	if ( /^@wordpress\//.test( request ) ) {
+		callback( null, formatRequest( request ), 'this' );
+	} else {
+		callback();
+	}
+};
+
+const externals = [
+	{
+		react: 'React',
+		'react-dom': 'ReactDOM',
+		tinymce: 'tinymce',
+		moment: 'moment',
+		jquery: 'jQuery',
+		lodash: 'lodash',
+		'lodash-es': 'lodash',
+	},
+	wordpressExternals,
+];
+
+const isProduction = process.env.NODE_ENV === 'production';
+const mode = isProduction ? 'production' : 'development';
+
+const config = {
+	mode,
+	entry: {
+		index: path.resolve( process.cwd(), 'src', 'index.js' ),
+	},
+	output: {
+		filename: '[name].js',
+		path: path.resolve( process.cwd(), 'build' ),
+	},
+	externals,
+	resolve: {
```

Comment:
Due to the way local packages work I can't test this properly with an external plugin. https://github.com/nosolosw/understanding-gutenberg/pull/6 doesn't work because `@wordpress/scripts` dependencies (being local) aren't hoisted to the top-level `node_modules` folder, but are installed in `node_modules/@wordpress/scripts/node_modules` instead. If I could use `@wordpress/scripts` from npm the dependencies would be hoisted and the example would work.

**There is a gotcha with using `pwd.cwd` here, though.**

As far as I understand, this tells webpack to lookup dependencies (`babel-loader`, `source-map-loader`) in the `node_modules` folder of the consumer (the plugin/app/etc that uses this config). Due to how npm works, we can't guarantee that the versions in the top-level `node_modules` of the consumer are the ones we expect. Example:

- a package other than `@wordpress/scripts` uses a different version of webpack/babel-loader/etc and it gets installed first. The top-level `node_modules/` will contain those versions.
- `@wordpress/scripts` is installed after, and because the top-level already has `webpack`, `babel-loader`, etc, its dependencies are installed under `node_module/@wordpress/scripts/node_modules`.

When webpack goes to find the dependencies, they won't be the ones we want. I don't know how frequent this can be. If the user only pulls `@wordpress/packages` as dependencies, this won't be a problem. But as soon as they start using external packages, it will.

**How to solve this?**

https://github.com/WordPress/gutenberg/pull/13877 is a potential approach. It doesn't use `pwd.cwd` but introduces some changes to path resolution and a plugin to teach Webpack how to behave.

We could use that as a base, merge, and then create a second PR on top of it that includes the changes in this PR.

Reply:
This is no longer a concern, as we figured we can use `require.resolve` for path resolution of the utils webpack uses. When the file is loaded, the paths will be resolved to point to `@wordpress/scripts`'s `node_modules` directory instead of the consumer.